driver-workload-executor/src/main/com/mongodb/workload/WorkloadExecutor.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.workload;
+
+import com.mongodb.WriteConcern;
+import com.mongodb.client.JsonPoweredCrudTestHelper;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonValue;
+import org.bson.json.JsonWriterSettings;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A workload executor for the driver.  It will execute the provided workload in a loop, until the process is terminated.
+ *
+ * The following system properties are supported:
+ *
+ *  org.mongodb.test.uri: the connection string required to connect to the cluster under test
+ *  org.mongodb.workload.spec: a JSON representation of the the workload
+ *  org.mongodb.workload.output.directory: the output directory in which to write the results.json
+ */
+public class WorkloadExecutor {
+    private static volatile boolean done;
+    private static final AtomicInteger numSuccesses = new AtomicInteger();
+    private static final AtomicInteger numFailures = new AtomicInteger();
+    private static final AtomicInteger numErrors = new AtomicInteger();
+
+    /**
+     * The main entry point
+     *
+     * @param args no arguments are expected
+     */
+    public static void main(String[] args) {
+        String outputDirectory = System.getProperty("org.mongodb.workload.output.directory") == null
+                ? System.getProperty("user.dir") : System.getProperty("org.mongodb.workload.output.directory");
+        
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+                done = true;
+                writeResults(outputDirectory);
```

Comment:
Write the results from the shutdown hook in order to avoid blocking process exit on hung I/O in the executor.